[Merged by Bors] - feat(Topology/CompletelyRegular): Add definition `CompletelyRegularSpace`

mapehe opened the pull request:
Add definitions `CompletelyRegularSpace` and `T35Space`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg commented:
I've taken the liberty to mark this as "topology". Hope that's fine with you!

mapehe commented:
For context, I'm trying to gradually build higher quality version of [this](https://github.com/leanprover-community/mathlib4/compare/master...mapehe:mathlib4:stone-cech-embedding#diff-2a6e4cb9e091ae059c58aee98c2b41490bf3da03d951473ce856c40a23363bffR601-R604) proof

alreadydone commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,111 @@
+@[mk_iff completelyRegularSpace_iff]
+class CompletelyRegularSpace (α : Type u) [TopologicalSpace α] [T1Space α] : Prop where
+  completely_regular :
+  ∀ (x : α), ∀ (K : Set α) (_: IsClosed K), Disjoint K {x} →
+    ∃ (f : C(α, ℝ)), EqOn f 0 {x} ∧ EqOn f 1 K ∧ ∀ x, f x ∈ Icc (0 : ℝ) 1
```
Maybe you're intentionally choosing your formulation but the following looks simpler:
```suggestion
  ∀ x : α, ∀ K : Set α, IsClosed K → x ∉ K →
    ∃ f : C(α, ℝ), f x = 0 ∧ EqOn f 1 K ∧ ∀ x, f x ∈ Icc (0 : ℝ) 1
```
Also note that `X Y Z` are preferred over `α γ` etc. if you're starting a new file.

alreadydone commented:
Note it's possible to define completely regular space using [existence of a compatible uniform structure](https://en.wikipedia.org/wiki/Tychonoff_space#Uniform_structures), and I think it would be preferred due to less imports. I don't know the proof of equivalence with the current definition, but I think it's something cool to add to mathlib.

mapehe commented:
> Note it's possible to define completely regular space using [existence of a compatible uniform structure](https://en.wikipedia.org/wiki/Tychonoff_space#Uniform_structures), and I think it would be preferred due to less imports (no mention of real-valued functions). I don't know the proof of equivalence with the current definition, but I think it's something cool to add to mathlib.

Thanks @alreadydone. I'll need to look into that.

alexjbest commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,111 @@
+        have T : f a = 0 := by
+          apply hf
+          rw [mem_singleton_iff]
+        exact Eq.mpr (id (T ▸ Eq.refl (f a ∈ u))) hhu
```
This looks like a bit too an extreme use of `show_term`, was it just a rw before or something? That's probably more readable

alexjbest commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,111 @@
+      exact Eq.mpr (id (propext disjoint_singleton_right ▸ Eq.refl (Disjoint s {a}))) ha
+    let h := CompletelyRegularSpace.completely_regular a s hs d
+    let f := h.choose
+    let ⟨hf, hhf, _⟩ := h.choose_spec
```
If this is a proposition it should be a  `have` not a `let`

alexjbest replied:
There is also a tactic `choose` that can make this sort of thing a bit less annoying to write

alreadydone replied:
Isn't this simply `obtain ⟨f, hf, hhf, _⟩ := CompletelyRegularSpace.completely_regular a s hs d`?
And if you don't want the `_` you can replace it by `-` so it doesn't show up in the context.

alexjbest commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,111 @@
+      have P : (Exists.choose h) x = 1 := hhf hx
+      simp only [eqOn_singleton, Pi.zero_apply, ge_iff_le, zero_le_one, not_true, gt_iff_lt,
+        mem_Icc] at P
+      rw [P]
+      exact hhv
```
You could use `rwa` (rewrite then assumption), here
```suggestion
      rwa [P]
```

alexjbest commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,111 @@
+    have duv : Disjoint (f ⁻¹' v) (f ⁻¹' u) := by
+      rw [Set.disjoint_iff, inter_comm, subset_empty_iff] at huv
+      rw [Set.disjoint_iff, ←preimage_inter, huv, preimage_empty]
+    exact
+      Exists.intro (↑f ⁻¹' v)
+        { left := mem_v, right := Exists.intro (↑f ⁻¹' u) { left := mem_u, right := duv } }
```
You might find the anonynous constructor notation helpful here, you can use `\<a, b\>` in place of `Exists.intro a b`, and you can nest this and replace `{left := a, right := b}` with the same thing too.

alreadydone replied:
i.e., `⟨f ⁻¹' v, mem_v, f ⁻¹' u, mem_u, duv⟩`

mapehe commented:
Thanks for the feedback on the code @alreadydone and @alexjbest. I made changes accordingly, happy to iterate, if you have further thoughts.

mapehe commented:
I feel like [this](https://github.com/leanprover-community/mathlib4/pull/7926/files#diff-bb6f0d47c3c6cb637ea412277814b30ac28d936daea70d1401bf6af6a256ba28R65-R68) should be easier, but I couldn't find a single `apply` for that. Maybe [these](https://github.com/leanprover-community/mathlib4/pull/7926/files#diff-bb6f0d47c3c6cb637ea412277814b30ac28d936daea70d1401bf6af6a256ba28R103-R110) two too.

mapehe commented:
> Note it's possible to define completely regular space using [existence of a compatible uniform structure](https://en.wikipedia.org/wiki/Tychonoff_space#Uniform_structures), and I think it would be preferred due to less imports (no mention of real-valued functions). I don't know the proof of equivalence with the current definition, but I think it's something cool to add to mathlib.

I thought about this a little bit. What I originally set out to work towards was
```
theorem denseEmbedding_stoneCechUnit [TopologicalSpace X] [T1Space X] [CompletelyRegularSpace X] :
    @DenseEmbedding _ _ _ _ (stoneCechUnit : X → StoneCech X)
```
The proof I currently have for that relies on [these](https://github.com/leanprover-community/mathlib4/pull/7926/files#diff-bb6f0d47c3c6cb637ea412277814b30ac28d936daea70d1401bf6af6a256ba28R56) functions in two ways

1.  Using them, you get a basis for the topology of `X`.
2. The codomain is a compact Hausdorff space. This enables the use of the universal property `stoneCechExtend_extends`.

I couldn't think of a simple way to understand the proof in terms of uniform structures, so, to me, it seems that this alternative definition would lead to a different direction with other possible goals. I guess it really comes down to what's considered interesting.

alreadydone commented:
I think it's fine to use the current definition for now, and we may or may not prefer to refactor it later using the uniformizable definition ... I think we could just add an instance `[UniformSpace X] : CompleteRegularSpace X` once the equivalence is proved.

mapehe commented:
> I think it's fine to use the current definition for now, and we may or may not prefer to refactor it later using the uniformizable definition ... I think we could just add an instance `[UniformSpace X] : CompleteRegularSpace X` once the equivalence is proved. I think this file will import real numbers anyway so my comment about imports is actually moot.

I think it may just be my lack of vision: As you mentioned above, it's possible to define these spaces and the claim I'm interested in without ever referring to reals. It doesn't seem (?) unreasonable to expect there would be some proof without reals too. Unfortunately that ventures to an area I don't know much about.

On the other hand, the present definition is a standard ingredient in the literature, so I'd expect it should be a part of mathlib in some form at some point. Perhaps there is an option to work with this definition for now, and improve on these issues later on? I was planning to add a couple of intermediary theorems next which are also standard stuff so perhaps there is recyclable material even if the definitions are rearranged in the future?

mapehe commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,109 @@
+instance CompletelyRegularSpace.t3space [CompletelyRegularSpace X] : T3Space X := by
+  have : T0Space X := by apply T1Space.t0Space
+  have : RegularSpace X := by
+    have ne : (0 : I) ≠ 1 := by
```
Is there a one-liner for this somewhere?

alreadydone replied:
[one_ne_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/NeZero.html#one_ne_zero).symm should work

mapehe commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,109 @@
+  let ⟨⟨f, cf⟩, hfx, hfK, hficc⟩ := exists_continuous_zero_one_of_closed cx hK d
+  let g : X → I := fun x => ⟨f x, hficc x⟩
+  have cg : Continuous g := cf.subtype_mk hficc
+  have hgx : g x = ⟨0, zero_mem⟩ := by
```
Is there a one-liner for this and `hgK`?

alreadydone replied:
Can't you simply write `g x = 0`? I think `Subtype.ext hfx` solves this.

alreadydone commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,104 @@
+    rw [mk_zero, ←coe_eq_zero]
+    apply hfx
+    rw [mem_singleton_iff]
+  have hgK : EqOn g 1 K := by
+    intro k
+    rw [Pi.one_apply, ←coe_eq_one]
+    apply hfK
```
Something like `fun k hk => Subtype.ext (hfK k hk)`
BTW you can Shift-select multiple lines when adding comments.

mapehe commented:
Thanks, was able to drop some boring lines based on your tips @alreadydone

alreadydone commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,93 @@
+/-- A T1 space X is completely regular if points can be separated from closed sets via
+  continuous functions to the unit interval. -/
+@[mk_iff completelyRegularSpace_iff]
+class CompletelyRegularSpace (X : Type u) [TopologicalSpace X] [T1Space X] : Prop where
+  completely_regular : ∀ (x : X), ∀ (K : Set X) (_: IsClosed K), x ∉ K →
+    ∃ f : X → I, Continuous f ∧ f x = 0 ∧ EqOn f 1 K
```
I think you should remove `T1Space` from the definition, because e.g. on [Wikipedia](https://en.wikipedia.org/wiki/Tychonoff_space) a completely regular space is not necessarily T1; if it's T0 then it's T3 and called a Tychonoff space. The equivalence with uniformizability only holds without the T1 condition, since a uniform space isn't necessarily T1.

Also I think you can use `C(X, I)` rather than separating out the `Continuous` condition. I know that @PatrickMassot urged you to separate it out, but I don't see the point: if you're getting out a continuous map from a `CompletelyRegularSpace`, then working with a bundled continuous map will only be convenient; if you're proving a `CompletelyRegularSpace`, then writing `⟨⟨f, cont_f⟩, ..⟩` instead of `⟨f, cont_f, ...⟩` isn't really much more difficult. I can see that an unbundled `Continuous` condition would be more convenient if it appears as an argument to a theorem, but not here. Note also that [Urysohn's lemma](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UrysohnsLemma.html#exists_continuous_zero_one_of_closed) is stated using bundled ContinuousMap.

mapehe replied:
About the terminology: I went back and forth a bit before making this PR, but eventually decided to use the definition given in Walker's book. Reconsidering this, I realize that some of the things I'm planning to add do not require the T1 bit, so it could be better to exclude T1 here, and maybe add a definition for another type of space (`Tychonoff`??) with T1 included.

About the bundled functions: I don't really have the sufficient background in contributing to mathlib to add any input on that.

alreadydone replied:
> it could be better to exclude T1 here, and maybe add a definition for another type of space (TychonoffSpace??) with T1 included.

Yeah it's definitely reasonable, e.g. [T3Space](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#T3Space) is T0 + RegularSpace. So CompleteRegularSpace without T0 is consistent with that, and CompleteRegularSpace + T0 could be called T35Space in analogy with [T25Space](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#T25Space).

PatrickMassot replied:
I just checked in Mathlib and *every* use of Urysohn's lemma unpack the bundle continuous map. So thanks @alreadydone for pointing out we should restate that lemma without bundling. I think it is not worth the trouble to fix Urysohnś lemma, but let's not reproduce that mistake here.

alreadydone replied:
Excuse me? I don't see it being unpacked at all in Metrizable, CompHaus/Basic, ContinuousFunction/Ideals, ContinuousMapDense among the [search results](https://github.com/search?q=repo%3Aleanprover-community%2Fmathlib4%20exists_continuous_zero_one_of_closed&type=code). Why would it be unpacked? ContinuousMap has a FunLike instance so if you want the function you can just pretend it's a function, and for the `Continuous` condition you can just write `f.continuous` or even `f.2`.

PatrickMassot replied:
Oh I see my mistake. I saw huge unpacking in lines like `let ⟨f, hfs, hft, hf⟩ := exists_continuous_zero_one_of_closed hs ht hd`, but you are right that the bundle continuous map is not unpacked. So there is indeed no need to fix those (but the also bring nothing).

PatrickMassot commented on Mathlib/Topology/CompletelyRegular.lean:
```diff
@@ -0,0 +1,93 @@
+  completely_regular : ∀ (x : X), ∀ (K : Set X) (_: IsClosed K), x ∉ K →
+    ∃ f : X → I, Continuous f ∧ f x = 0 ∧ EqOn f 1 K
+
+instance CompletelyRegularSpace.t3space [CompletelyRegularSpace X] : T3Space X := by
```
This proof is painful because you do not let filters help you. Whenever you need to discuss elements of a filter, you are either building the early filter library or fighting it. Here you are fighting it.

Once you have your function `f` sending `a` to `0` and `s` to 1, the intuitive proof is the following: 
* `f` sends the "set" of points that are close to `a` into the "set" of points that are close to `0` (by continuity of `f` and assumption on `f a`).
* `f` sends the "set" of points that are near `s` into the "set" of points that are close to `1` (by continuity of `f` and assumption on `f '' s`).
* Because `I` is `T2` and `0 ≠ 1`, those target "sets" are disjoints, hence the sources "sets" are disjoint.

The above arguments uses filters on `X` as generalized sets in `X`, it never uses a set of `X` seen as a member of a filter.

Unfortunately two lemmas seem to be missing from Mathlib. The first one is very surprising. It covers the "hence" in the last sentence above. Mathlib seems to have only the iff version which requires an injectivity assumption. Note the proof is fully obfuscated since the statement is obvious.

```lean
-- This lemma needs to go in `Order.Filter.Basic`, before `disjoint_map`
lemma Filter.disjoint_of_map {α β : Type*} {F G : Filter α} {f : α → β}
    (h : Disjoint (map f F) (map f G)) : Disjoint F G :=
    disjoint_iff.mpr <| map_eq_bot_iff.mp <| le_bot_iff.mp <| trans map_inf_le (disjoint_iff.mp h)
```

Then it  is slightly more convenient to add a special case of a lemma we already have:
```lean
-- This lemma needs to go in `Topology.NhdsSet`, after `Continuous.tendsto_nhdsSet`.
lemma Continuous.tendsto_nhdsSet_nhds {X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y]
    {s : Set X} {y : Y} {f : X → Y} (h : Continuous f) (h' : EqOn f (fun _ ↦ y) s) : 
    Tendsto f (𝓝ˢ s) (𝓝 y) := by 
  rw [← nhdsSet_singleton]
  exact h.tendsto_nhdsSet h'
```
(needing to add stuff there is a lot less surprising since `nhdsSet` isn't very much used).

Now you need to pay for your idea to use `I` as the target instead of `ℝ`. It seems Mathlib doesn't know that `0 ≠ 1` in this type.
```
instance : NeZero (1 : I) := ⟨sorry⟩
```

And then you are ready to follow the intuitive sketch above:
```lean
instance CompletelyRegularSpace.t3space [CompletelyRegularSpace X] : T3Space X := by
  have : T0Space X := by apply T1Space.t0Space
  have : RegularSpace X := (regularSpace_iff X).mpr <| by    
    intro s a hs ha
    obtain ⟨f, cf, hf, hhf⟩ := CompletelyRegularSpace.completely_regular a s hs ha
    apply disjoint_of_map (f := f)
    apply Disjoint.mono (cf.tendsto_nhdsSet_nhds hhf) cf.continuousAt
    exact disjoint_nhds_nhds.mpr (hf.symm ▸ zero_ne_one).symm
  exact {}
```

alreadydone replied:
Threre is [unitInterval.instNeZeroElemRealUnitIntervalHasZeroOfNatToOfNat1HasOne](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UnitInterval.html#unitInterval.instNeZeroElemRealUnitIntervalHasZeroOfNatToOfNat1HasOne) and I'm adding ZeroLEOne and BoundedOrder [here](https://github.com/leanprover-community/mathlib4/pull/7915/files#diff-fa7d5572056db41c1badb453962d73e25039fdaab5aa06bbd70db9221b99727fR75-R77).

mapehe replied:
Thanks, this simpler proof will definitely be useful as a reference for me.

I added the two theorems to the designated places. Someone seems to have added `NeZero` in master recently, so no need to add anything there probably.

mapehe commented:
@alreadydone I switched to separate definitions of `CompletelyRegularSpace` and `T35Space` and adjusted the theorems about which space types imply others accordingly. I think the current terminology indeed fits much better with the existing space types.

PatrickMassot commented:
bors merge

PatrickMassot commented:
Thanks @mapehe, and congratulations for your first accepted PR to Mathlib!
